app/src/main/java/com/example/anfcodetest/MainActivity.kt

```diff
@@ -1,55 +1,36 @@
 package com.example.anfcodetest
 
-import androidx.appcompat.app.AppCompatActivity
 import android.os.Bundle
 import androidx.recyclerview.widget.LinearLayoutManager
-import com.fasterxml.jackson.module.kotlin.*
+import com.hannesdorfmann.mosby3.mvp.MvpActivity
 import kotlinx.android.synthetic.main.activity_main.*
-import okhttp3.*
-import java.io.IOException
 
-val url = "https://www.abercrombie.com/anf/nativeapp/qa/codetest/codeTest_exploreData.json"
-
-class MainActivity : AppCompatActivity(), MainContract.View {
+class MainActivity : MvpActivity<MainContract.View, MainContract.Presenter>(), MainContract.View {
 
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
     setContentView(R.layout.activity_main)
-
     loadCards()
   }
 
+  override fun displayCards(cards: List<Card>) {
+    val adapter = MainAdapter(cards)
+    runOnUiThread {
+      recyclerView_main.adapter = adapter
+      adapter.notifyDataSetChanged()
+    }
+  }
+
   override fun loadCards() {
-    fetchJson(url)
+    getPresenter().fetchJson(PAGE_DATA_URL)
     recyclerView_main.layoutManager = LinearLayoutManager(this)
     recyclerView_main.recycledViewPool.setMaxRecycledViews(0,0)
+    recyclerView_main.adapter = MainAdapter(emptyList())
```

Comment:
Here we can set an empty adapter and just wait for our data to come back, this way you can support async loading like you have for the JSON! Up above, on the callback, we can reload our views with our new data.